source/server/listener_manager_impl.cc

```diff
@@ -669,18 +673,28 @@ bool ListenerManagerImpl::removeListener(const std::string& name) {
 void ListenerManagerImpl::startWorkers(GuardDog& guard_dog) {
   ENVOY_LOG(info, "all dependencies initialized. starting workers");
   ASSERT(!workers_started_);
-  workers_started_ = true;
   uint32_t i = 0;
+  const auto listeners_pending_init =
+      std::make_shared<std::atomic<uint64_t>>(workers_.size() * active_listeners_.size());
   for (const auto& worker : workers_) {
     ASSERT(warming_listeners_.empty());
     for (const auto& listener : active_listeners_) {
-      addListenerToWorker(*worker, *listener);
+      addListenerToWorker(*worker, *listener, [this, listeners_pending_init]() {
+        if (--(*listeners_pending_init) == 0) {
+          workers_started_ = true;
```

Comment:
I have moved `workers_started_` variable also here to truly reflect its status. If there are any objections/risks, I can revert that change.

Reply:
Looks `workers_started_` has some side affects, especially when trying to `stopWorkers` immediately before it is actually set to true - so reverted this change.